### ISSUE TITLE
FIFOQueue: dequeue many operation very slow?

### ISSUE THREAD

**Issue by tomrunia** [tensorflow/tensorflow#3009]
When training a relatively simple model (1-layer LSTM, 256 units) my Titan X GPU keeps spiking from 0% to 30% GPU utilization. Conclusion: somewhere in the pipeline there is a bottleneck which limits the GPU to be processing the training batches continuously. I use a FIFOQueue to which examples are being fed in one or more separate threads:

```python
queue = tf.FIFOQueue(
     capacity=self.config.queue_capacity,
     dtypes=[tf.float32, tf.float32],
     shapes=[[30, 49, 512], [30]],
     name="FIFOQueue"
)
```

For the training operation I use `queue.dequeue_many` to get examples from the queue. As you can see the batch size is 64 examples. So in the end the input tensor is `64x30x49x512` of type `tf.float32`:

```python
# Model inputs, either use the queue (training) or feed_dict (evaluation)
inputs, targets = queue.dequeue_many(64)
```

To find out why my code is running "slow" (i.e. spiking GPU allocation and no temperature increase) I use the `Timeline` object ([see here](http://stackoverflow.com/questions/34293714/tensorflow-can-i-measure-the-execution-time-of-individual-operations)) to measure execution times of individual operations. The results displayed below show the measurements for one training iteration at which point the queue was filled with more than 1000 examples. I have included screenshots for both GPU and CPU-only runs (forced with `export CUDA_VISIBLE_DEVICES=-1`. 

What strikes me from these results is that it takes a really long time to dequeue examples from the FIFOQueue. What is happening here...something wrong or is the dequeuing operation just very slow? Overall the dequeuing operation and sending the data to the GPU takes up half of the time of a training iteration. No wonder that the GPU utilization is spiking. Any help is welcome optimizing my training pipeline! As I understand correctly the examples are all queued in RAM, is there also a way to queue them ahead on GPU memory so when they are needed they do not have to be moved CPU => GPU?

This is tested on TensorFlow v9.0 build from sources about 1.5 week ago.

**GPU running on Titan X**
![gpu](http://i.imgur.com/eTZNPDM.png?1) 

**CPU running on Xeon CPU E5-2640**
![cpu](http://i.imgur.com/mQkLH8h.png?1)

**Comment by prb12** [comment]
Thanks for including the timelines!  Nice to see people using that ;-)

It's hard to say very much about what's happening here without seeing the whole program.  I don't _believe_ that dequeueing 192MB is inherently that expensive.  (@mrry - does this sound right?)

There are a number of things which might be happening.  Obviously `DequeueMany` is a blocking operation and the time it appears to take in the timeline will include all the time from when it was issued until when the final element was enqueued into the FIFO.  This timeline will not show the other, potentially concurrent steps which are happening.

Are you enqueueing all 64 required elements before calling the step which dequeues them? 
Can you contrive a test case where you put 64 elements into the queue in advance, and just run a step which does the dequeue?

**Comment by mrry** [comment]
@prb12: At the very least, the DequeueMany op is going to do 64 separate Eigen slice assignments to build up the output batch. If the capacity is less than 64, it will also have the copies out of any corresponding EnqueueMany op arguments accounted to it. The most likely cause is that it's blocking, however, so it would be great if @tomrunia could do the contrived test that you're suggesting. Otherwise using a sampling profiler might reveal a suprisingly costly copy.

**Comment by tomrunia** [comment]
Oke, I wrote a simple script (see below) in which the threads simple enqueue batches of `np.ones([64, 30, 49, 512], dtype=np.float32)` to the `FIFOQueue`. The main loop simple dequeues 64 examples and performs a simple `tf.square` of the input tensor. Before the main loop starts I make sure that the queue is filled with a sufficient number of examples. Again, the dequeue operation takes a "long" time to finish as you can see in the timelines below. 

**GPU running on Titan X**
![gpu](http://i.imgur.com/VrKa51G.png)

**CPU running on Xeon E5-2640**
![cpu](http://i.imgur.com/VlG68ZX.png)

```python
import time
import numpy as np
import threading

import tensorflow as tf
from tensorflow.python.client import timeline


def test_queue():

    feature_input = tf.placeholder(tf.float32, shape=[None, 30, 49, 512], name="queue_inputs")
    target_intput = tf.placeholder(tf.float32, shape=[None, 30], name="queue_targets")

    queue = tf.FIFOQueue(
        capacity=1000,
        dtypes=[tf.float32, tf.float32],
        shapes=[[30, 49, 512], [30]],
        name="FIFOQueue"
    )

    # Enqueue and dequeue operations
    enqueue_op = queue.enqueue_many([feature_input, target_intput])
    queue_inputs, queue_targets = queue.dequeue_many(64)

    queue_size = queue.size()

    # Very simple training operator, dequeue examples from the FIFOQueue and
    # square the matrix elementwise. Just for testing of course.
    square_op = tf.square(queue_inputs)

    # Coordinator for threads
    coord = tf.train.Coordinator()

    # Initialize the TensorFlow session
    gpu_options = tf.GPUOptions(
        per_process_gpu_memory_fraction=0.75,
    )
    
    sess = tf.Session(config=tf.ConfigProto(
        gpu_options=gpu_options,
        log_device_placement=False,
        allow_soft_placement=False
    ))

    def load_and_enqueue():
        while True:
            # Just feed random stuff to the queue
            #features = np.random.rand(64, 30, 49, 512)
            features = np.ones([64, 30, 49, 512], dtype=np.float32)
            targets  = np.ones([64, 30], dtype=np.float32)

            # Feed example to Tensorflow placeholder
            feed_dict = {
                feature_input: features,
                target_intput: targets
            }

            # Push all the training examples to the queue
            sess.run(enqueue_op, feed_dict=feed_dict)

            if coord.should_stop():
                break

    # Start the threads
    num_threads = 4
    for i in range(num_threads):
        t = threading.Thread(target=load_and_enqueue)
        t.setDaemon(True)
        t.start()

    # Make sure the queueu is filled with some examples (n = 500)
    num_samples_in_queue = 0
    while num_samples_in_queue < 500:
        num_samples_in_queue = sess.run(queue_size)
        print("Initializing queue, current size = %i" % num_samples_in_queue)
        time.sleep(1)

    # Initialize the session
    init = tf.initialize_all_variables()
    sess.run(init)

    run_options  = tf.RunOptions(trace_level=tf.RunOptions.FULL_TRACE)
    run_metadata = tf.RunMetadata()

    for step in range(1000):

        print("Step = %i, QueueSize = %i" % (step, sess.run(queue_size)))

        if step == 10:
            # Perform a step with saving the results as timeline object
            result = sess.run(square_op, options=run_options, run_metadata=run_metadata)

            # Create the Timeline object, and write it to a json
            tl = timeline.Timeline(run_metadata.step_stats)
            ctf = tl.generate_chrome_trace_format()
            with open('timeline.json', 'w') as f:
                print("writing to timeline.json")
                f.write(ctf)

        else:
            # Perform a step without saving the results
            result = sess.run(square_op)

    # Ask for the threads to stop
    coord.request_stop()
    sess.close()


if __name__ == "__main__":
    test_queue()

```

**Comment by prb12** [comment]
@tomrunia Thanks for the clear and self-contained repro code. It really helps us to have good quality reports for performance issues like this (and I especially like having the timelines!) 
 
We'll take a look into this.

**Comment by prb12** [comment]
I spent some time looking into what's going on here...   

As I mentioned earlier, your input batch of  [64, 30, 49, 512] * tf.float32 equates to 192MB...  

First, I established a baseline cost for memcpying a tensor of the size you are feeding.  A native C++ program calling memcpy in a tight loop between buffers of this size takes around **22ms** per iteration (single threaded).  The TensorFlow enqueue/dequeue operations actually use the _Eigen_ library to copy out slices of the tensors and this is likely to be **less** efficient than a flat memcpy since it needs to handle general shapes of n-D arrays.   Theoretically Eigen might be able to do this in parallel, but doesn't appear to be in this case (from looking at the CPU usage).

Because you are feeding in numpy arrays, and retrieving a large result, you incur a copy in each direction within the session.Run() call,  [here](https://github.com/tensorflow/tensorflow/blob/dfb71ea206eb9f61e5d97c9727caa1a6449e39cb/tensorflow/python/client/tf_session_helper.cc#L445) and [here](https://github.com/tensorflow/tensorflow/blob/dfb71ea206eb9f61e5d97c9727caa1a6449e39cb/tensorflow/python/client/tf_session_helper.cc#L525) for TensorFlow 0.9.  

Note - this code path is slightly different in various versions of TensorFlow.  Also, if you ever move to the distributed runtime these 192MB tensors will get serialized as ProtoBufs and this is *very* expensive.

Also, as @mmry says, some of the CPU time related to enqueue operations is most likely being accounted to the DequeueMany op due to [this](https://github.com/tensorflow/tensorflow/blob/r0.9/tensorflow/core/kernels/queue_base.cc#L336) code - once the queue is full, enqueue ops are blocked and wait in a list.  They can then get executed the next time a dequeue succeeds.

Note that a queue with max size of 1000 elements is about **3GB**, so depending on your machine config you may be causing a lot of virtual memory (and allocator) pressure.  This _isn't_ happening on my machine, but you may want to check by running `htop` or `vmstat` while executing your program.  A smaller queue may be more sensible.

In general, you may be better off using one of the TensorFlow input ops which reads image data directly from the file sytem, and then do any preprocessing as part of the graph.  (as opposed to preprocessing in Python and feeding in the raw tensor data).

For reference, I've attached a [pprof](http://goog-perftools.sourceforge.net/doc/cpu_profiler.html) profile of your code (with a few relatively insignificant changes).  You can see that the bulk of the cycles are spent in `TF_Run_wrapper_helper` (doing memcpys of the numpy arrays), and under `QueueBase::FlushUnlocked` doing both the enqueue and dequeue copy ops via _Eigen_ (one of which ends up turning into a `memcpy`)

Hope this helps .... 
Paul

![profile](https://cloud.githubusercontent.com/assets/11547801/16394032/140cbc64-3c68-11e6-8fe3-6335de4ab229.png)

**Comment by tomrunia** [comment]
Hi Paul, thanks a lot for your incredibly detailed analysis of what's happening here. This give me much more insight in what's is happening under the hood. I will rethink about the data processing (input) pipeline and see if I can optimize it here and there and decrease the queue size. Great work with the analysis!

**Comment by rohitgirdhar** [comment]
I was facing similar issues, copying large amounts of data from numpy to TF. I moved all the data loading and preprocessing into the graph (using https://www.tensorflow.org/versions/r0.9/how_tos/reading_data/index.html) and jpeg_decode, resize_images etc functionalities, and it sped up the training significantly.

**Comment by sjperkins** [comment]
I've noticed memcpy's being mentioned in this and [other](https://github.com/tensorflow/tensorflow/issues/2919) issues related to performance. I'm guessing memcpy is used to avoid thread-safety issues, in particular to support operations that take references to tensors such as **assign**.

Has any thought been given to introducing Copy-On-Write functionality to tensors within the graph? Numpy handles [this](http://docs.scipy.org/doc/numpy/reference/generated/numpy.ndarray.flags.html#numpy.ndarray.flags) but obviously without the tricky thread safety issue.

I ask because it seems wasteful in terms of I/O and memory to spend a memcpy feeding a numpy array onto a FIFOQueue and another memcpy (within a graph) copying the resultant tensor off the FIFOQueue. Ideally it would be nice to use FIFOQueue as a a pure synchronisation construct ala. go channels, python Queues rather a sychronised buffer. Clearly this needs to happen in cases where tensors cross device boundaries (GPU's/Network interfaces) but might there be scope for performance gains here?

I also notice that memcpy's happen within slicing operations. Could this be replaced with a Eigen::TensorMap (homologous to a numpy view) for e.g.?

**Comment by aselle** [comment]
I think @prb12 's really detailed analysis answers the question. There is always a tradeoff in using convenient and general operations for data manipulation as well as using numpy source data. We are continually working to improve performance. Automatic fusion of ops as alluded in the original paper's future work and issue #164 will provide generality and good performance.  

Closing for now.

**Comment by rohitgirdhar** [comment]
@prb12 Thanks for your detailed analysis! I am facing similar issues (timeline attached). I'm using tensorflow read operations throughout, so I believe my problem is not the numpy<->TF data copy. However your comment on virtual memory/allocator pressure seems related to my problem. Here's a screenshot from `htop`, and also the `vmstat` output
```text
procs -----------memory---------- ---swap-- -----io---- -system-- ------cpu-----
 r  b   swpd   free   buff  cache   si   so    bi    bo   in   cs us sy id wa st
 4  0 9744848 1865340 2888112 45542040    1    8   118    50    0    0 21 18 61  0  0
```
![screen shot 2016-07-25 at 1 24 57 pm](https://cloud.githubusercontent.com/assets/1893429/17116146/46c40090-526b-11e6-8071-ef45b0fdc881.png)
It seems my virtual memory use is 160G, do you think that might be the problem? I am not really familiar with these system level details, so I would greatly appreciate if you could elaborate on this problem. Thank you!

![screen shot 2016-07-25 at 10 55 04 am](https://cloud.githubusercontent.com/assets/1893429/17116222/a385f1f8-526b-11e6-846f-53ce2140edf1.png)

**Comment by rohitgirdhar** [comment]
I have an additional observation that might help us debug this issue. The slowdown for me happens when my cache memory gets full. I tried clearing the cache using:
```bash
$ sync && sudo echo 3 > /proc/sys/vm/drop_caches
```
The training then runs perfectly until the cache gets ~3/4 full again (as seen by the yellow bars on `htop`), and then it gets slow. So far my solution is to clear cache and restart from the last snapshot. Hoping there's a better way to fix this!

**Comment by Jerryzcn** [comment]
@rohitgirdhar clearing cache works for me, when i was training inception v3.

**Comment by prb12** [comment]
@rohitgirdhar I'm not 100% sure I want to add to an closed issue, commenting on something which may not even be related to the original problem, but...

From the `htop` output and your `drop_caches` observation it sounds to me like there may be some memory pressure caused by virtual address space fragmentation and high system buffer cache churn (reading large training datasets from the file system).  Can you see if your program is causing lots of page faults?  e.g. [like this](http://www.cyberciti.biz/faq/linux-command-to-see-major-minor-pagefaults/)

Internally at google we almost always use the [TCMalloc](http://goog-perftools.sourceforge.net/doc/tcmalloc.html) memory allocator which appears to be much better for TensorFlow workloads (and most others at google!).  When running the open source TensorFlow build via a standard Python binary the only way to use TCMalloc would be to inject a different heap implementation into Python using `LD_PRELOAD`, e.g.
`LD_PRELOAD="/usr/lib/libtcmalloc.so" python myprogram.py`

It would be very useful if you could try running with TCMalloc and see if it makes any difference.

**Comment by Jerryzcn** [comment]
The relevant issue is https://github.com/tensorflow/models/issues/170

**Comment by rohitgirdhar** [comment]
I think TCMalloc was the solution for me, as I no longer need to clear the cache to maintain the same training speed. Thanks @prb12 !

**Comment by lightcatcher** [comment]
Is there any progress on improving Numpy -> GPU throughput with Tensorflow?
 
I have a similar benchmark to @tomrunia of passing a O(100MB) tensor from Numpy to GPU and returning a small slice of the tensor. This would ideally be bottlenecked by PCI-e bandwidth, but runs at ~1/10th of that rate. Additionally, I see very poor scaling when attempting to run this tasks over multiple GPUs (on a single node), while theoretically I should benefit from the additional PCI-e bandwidth. (all stats from Tensorflow 0.9 built from source).

@prb12's description is very useful, but I'd like to understand more of what's going on when passing through feed_dict, enqueuing, dequeuing.
My understanding:
When a Numpy tensor is passed through a feed_dict it is copied into a host side Eigen tensor.
When a tensor is enqueued, it is added to a queue that lives on the host.
When a tensor is dequeued on the GPU, a cudaMemcpy copies the tensor from host to device. This should run near PCI-e bandwidth.

Questions about my understanding:
Is enqueueing a tensor done by reference or with a full copy (and memory allocation)?
Are the memcpy's on feed_dict passing (and enqueue if there's a memcpy there) performed by a single thread for the full Tensorflow session? Even if the ``Session.run`` calls come from multiple Python threads?

Are there plans to fix this, such as a queue that is resident on the GPU?

**Comment by mkolod** [comment]
BTW, I tried TCMalloc on AlexNet with 1-4 Pascal-grade GPUs (Titan X, P100). The queues are doing the threading for the CPU-side JPEG decoding, and since GPUs have very high throughput, threading of the CPU-side code is critical here. TCMalloc does speed things up by 20% for 1-4 threads (inter/intra threads, threads associated with custom queues, etc.), but there's a break-even point at 8 threads, and beyond that TCMalloc does more harm than benefit. At 20 threads (e,g, on a 20-core Intel Xeon E7-8870 v4), perf. is actually 40% down by using TCMalloc in such a heavily threaded environment. So, I doubt that TCMalloc is a universally beneficial solution,

**Comment by variac** [comment]
TCMalloc also seems to suffer from page faults, but it deals with them better for some reason. You can prominently see page fault activity in the kernel using `perf top -p <PID> -g`. Jemalloc gets completely horked on this when training on large datasets like ImageNet, though. 50+% drop in performance, and never recovers on its own. This is _really bad_ user experience. I would wager most TF users don't have the skills to diagnose or fix this. I'd much prefer something as straightforward as a data loader would work out of the box.

**Comment by agupta74** [comment]
As mentioned by @prb12 in the comment dated June 27, 2016,  based on my Cuda profiles using nvprof, I also see that that host to device transfers are not pinned in Tensor Flow v 1.2.1. Any ideas on when this will be fixed?

**Comment by yaroslavvb** [comment]
it's probably better off opening a separate issue to address the "not pinned" problem and link to relevant info, so that it could get triaged properly

**Comment by hkxIron** [comment]
@rohitgirdhar ,could you paste your code about function _read_image_from_disk()? I have verify that parse_example is much faster than parse_single_example.
